File: tools/ghidra/apply_labels_headless.py
# Apply labels to MicroBrute firmware - headless version
# Usage: Ghidra/support/analyzeHeadless <project-dir> <repo-name> -import <firmware-file> -scriptPath <path-to-this-script> -postScript apply_labels_headless.py

import re
from ghidra.app.script import GhidraScript
from ghidra.program.model.address import *

class ApplyLabelsHeadless(GhidraScript):
    def run(self):
        header_path = "/Users/jordan_after_midnight/Projects/music/macrobrute/firmware/ghidra_decompiled.h"
        
        self.print("Reading header file: " + header_path)
        
        with open(header_path, 'r') as f:
            content = f.read()

        # Pattern to match function declarations
        pattern = r'(thunk_)?FUN_([0-9a-fA-F]{8})\s*\('
        matches = re.findall(pattern, content)
        
        self.print("Found %d function names" % len(matches))

        seen = set()
        count = 0
        failed = 0
        
        for prefix, addr in matches:
            addr_int = int(addr, 16)
            if addr_int not in seen:
                seen.add(addr_int)
                name = "thunk_FUN_%s" % addr if prefix else "FUN_%s" % addr
                try:
                    addr_obj = toAddr(addr_int)
                    func = getFunctionAt(addr_obj)
                    if func:
                        func.setName(name)
                        count += 1
                    else:
                        # Try to create function if it doesn't exist
                        func = createFunction(addr_obj, name)
                        if func:
                            count += 1
                        else:
                            failed += 1
                except Exception as e:
                    failed += 1

        self.print("Applied %d labels!" % count)
        self.print("Failed or not found: %d" % failed)

File: tools/ghidra/ApplyLabelsHeadless.java
// Apply labels to MicroBrute firmware - Java version
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.SourceType;
import java.util.regex.*;

public class ApplyLabelsHeadless extends GhidraScript {
    public void run() throws Exception {
        String headerPath = "/Users/jordan_after_midnight/Projects/music/macrobrute/firmware/ghidra_decompiled.h";
        
        println("Reading header file: " + headerPath);
        
        java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.FileReader(headerPath));
        StringBuilder content = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            content.append(line).append("\n");
        }
        reader.close();
        
        Pattern pattern = Pattern.compile("(thunk_)?FUN_([0-9a-fA-F]{8})\\s*\\(");
        Matcher matcher = pattern.matcher(content.toString());
        
        java.util.Set<Integer> seen = new java.util.HashSet<>();
        int count = 0;
        int failed = 0;
        
        while (matcher.find()) {
            String prefix = matcher.group(1);
            String addrStr = matcher.group(2);
            int addrInt = Integer.parseInt(addrStr, 16);
            
            if (!seen.contains(addrInt)) {
                seen.add(addrInt);
                String name = (prefix != null ? "thunk_FUN_" : "FUN_") + addrStr;
                
                try {
                    Address addrObj = toAddr(addrInt);
                    Function func = getFunctionAt(addrObj);
                    if (func != null) {
                        func.setName(name, SourceType.USER_DEFINED);
                        count++;
                    } else {
                        func = createFunction(addrObj, name);
                        if (func != null) {
                            count++;
                        } else {
                            failed++;
                        }
                    }
                } catch (Exception e) {
                    failed++;
                }
            }
        }
        
        println("Found " + seen.size() + " unique function names");
        println("Applied " + count + " labels!");
        println("Failed or not found: " + failed);
    }
}